refactor/admin: Admin Verb Datums

littleboobs opened the pull request:
## Что этот ПР делает
- https://github.com/tgstation/tgstation/pull/82511
- https://github.com/ParadiseSS13/Paradise/pull/30646
- https://github.com/Monkestation/Monkestation2.0/pull/5944

В указанных пр-ах написаны умные слова про удобство, инкапсуляцию пермишенов и бла-бла-бла, можете почитать. Портирован вариант ТГ. Сами вёрбы почти не трогались.

Единственное, что может смущать это перемещение некоторых вербов из одной категории в другую, а также добавление некоторых новых вербов в контекстное меню. Возможно, что некоторые вербы не работают, для этого и нужен ТМ. Не надо трястись и тратить лишнюю энергию, просто сообщите мне в баг-репорте.

## Тестирование
Протыкал большинство вёрбов и надеюсь, что отловил все косяки, но нужен ТМ.

littleboobs committed:
base

datum, subsystem, defines and co.

littleboobs committed:
rename defines and split

littleboobs committed:
[1] BIG COMMIT

littleboobs committed:
Merge remote-tracking branch 'upstream/master220' into admin-verbs

littleboobs committed:
[2] BIG COMMIT

littleboobs committed:
statpanel -> verb_category

dageavtobusnick commented on code/__HELPERS/mobs.dm:
```diff
@@ -900,3 +900,25 @@ GLOBAL_DATUM_INIT(dview_mob, /mob/dview, new)
+		//	var/mob/living/carbon/carbon_target = mob_target
+		//	if(!QDELETED(carbon_target.last_mind))
+		//		return carbon_target.last_mind
+	else if(istype(target, /obj/item/organ/internal/brain))
```
Ретурны заместо else

dageavtobusnick replied:
изхелпер. Если нет, то сделать

dageavtobusnick commented on code/modules/vote/vote_verb.dm:
```diff
@@ -13,65 +13,55 @@
-
-	if(!check_rights(R_ADMIN))
+ADMIN_VERB(start_vote, R_ADMIN, "Start Vote", "Start a vote on the server.", ADMIN_CATEGORY_MAIN)
+	if(SSvote.active_vote)
```
Это по идее у нас не нужно. У нас воуты помещаются в очередь.

littleboobs replied:
убрал

dageavtobusnick commented on code/modules/shuttle/emergency.dm:
```diff
@@ -97,7 +97,6 @@
 	var/sound_played = 0 //If the launch sound has been sent to all players on the shuttle itself
 
 	var/canRecall = TRUE //no bad condom, do not recall the crew transfer shuttle!
-	var/forceHijacked = FALSE // forced change of arrival at the syndicate base
```
А зачем ты это убрал?

littleboobs replied:
Уже не помню, точно помню, что хотел спросить об этом вербе. Посмотрю, верну

dageavtobusnick commented on code/controllers/subsystem/non-firing/admin_verbs.dm:
```diff
@@ -0,0 +1,153 @@
+		if(verb_singleton.visibility_flag)
+			if(!(verb_singleton.visibility_flag in admin_verbs_by_visibility_flag))
+				admin_verbs_by_visibility_flag[verb_singleton.visibility_flag] = list()
+			admin_verbs_by_visibility_flag[verb_singleton.visibility_flag] |= list(verb_singleton)
```
ранний continue

dageavtobusnick commented on code/controllers/subsystem/non-firing/admin_verbs.dm:
```diff
@@ -0,0 +1,153 @@
+		else for(var/permission_flag in bitfield_to_list(verb_permissions))
+			if(!has_permission["[permission_flag]"])
+				continue
+			valid_verbs |= list(verb_singleton)
```
ранний continue заместо else

dageavtobusnick commented on code/controllers/subsystem/non-firing/admin_verbs.dm:
```diff
@@ -0,0 +1,153 @@
+	UnregisterSignal(admin, COMSIG_CLIENT_MOB_LOGIN)
+	for(var/datum/admin_verb/verb_type as anything in admin_verbs_by_type)
+		admin_verbs_by_type[verb_type].unassign_from_client(admin)
+	admin_visibility_flags -= list(admin.ckey)
```
во всем файле отдели if, while, for от остального кода пустыми строками

dageavtobusnick commented on code/game/jobs/job_exp.dm:
```diff
@@ -46,22 +46,21 @@ GLOBAL_LIST_INIT(role_playtime_requirements, list(
+/// Allows admins to determine who the newer players are.
+ADMIN_VERB(check_player_exp, R_ADMIN|R_MOD|R_MENTOR, "Check Player Playtime", "Return a playtime report.", ADMIN_CATEGORY_MAIN)
+	var/list/msg = list()
+	msg += "<html><meta charset='utf-8'><head><title>Playtime Report</title></head><body>"
```
хедер тут убран не по приколу. Попап его предоставляет сам по себе. Если где-то еще такое делал - убери. Везде где надо хедеры стоят.

littleboobs replied:
Окей, вроде бы только тут добавлял

dageavtobusnick commented on code/game/objects/explosion.dm:
```diff
@@ -86,8 +83,11 @@
+
+/proc/wipe_color_and_text(list/atom/wiping)
+	for(var/i in wiping)
+		var/atom/A = i
```
имя переменной + as anything внутри for, заместо приведения тут.

dageavtobusnick commented on code/modules/admin/admin_investigate.dm:
```diff
@@ -1,17 +1,10 @@
-	if(!check_rights(R_ADMIN))
-		return
+	var/file = wrap_file("[GLOB.log_directory]/[subject].html")
+	WRITE_FILE(file, "[time_stamp()] [UID()] [ADMIN_COORDJMP(src)] || [src] [message]<br>")
```
Просто тут лог в файл, а не админу, зачем тут jmp?. Бери с тг тогда уж https://github.com/tgstation/tgstation/blob/37dc763ff9d948db4532eb346c38712fc6f00ff0/code/modules/admin/admin_investigate.dm#L13

littleboobs replied:
Странно, как будто с тг и брал)) Сделаю

dageavtobusnick commented on code/modules/admin/admin_verbs.dm:
```diff
@@ -520,98 +118,78 @@ GLOBAL_LIST_INIT(view_runtimes_verbs, list(
+			explosion(epicenter, devastation_range = 1, heavy_impact_range = 2, light_impact_range = 3, flash_range = 3, cause = user.mob)
+		if("Средняя бомба (2, 3, 4, 4)")
+			explosion(epicenter, devastation_range = 2, heavy_impact_range = 3, light_impact_range = 4, flash_range = 4, cause = user.mob)
+		if("Большая бомба (3, 5, 7, 5)")
```
Дефайнизируй выборы

dageavtobusnick commented on code/modules/admin/admin_verbs.dm:
```diff
@@ -756,294 +327,115 @@ GLOBAL_LIST_INIT(view_runtimes_verbs, list(
+	var/message = tgui_input_text(user, "What do you want the message to be?", "Make Sound")
+	if(!message)
 		return
+	for(var/mob/V in hearers(O))
```
имя переменной

dageavtobusnick commented on code/modules/admin/admin_verbs.dm:
```diff
@@ -756,294 +327,115 @@ GLOBAL_LIST_INIT(view_runtimes_verbs, list(
-	set desc = "Display a message to everyone who can hear the target"
-
-	if(!check_rights(R_SOUNDS))
+ADMIN_VERB_ONLY_CONTEXT_MENU(make_sound, R_SOUNDS, "Make Sound", obj/O in view())
```
Имя переменной во всех подобных местах

littleboobs replied:
Да не хотелось трогать вербы, чтобы диф не тащить, ну лан

dageavtobusnick commented on code/modules/admin/admin_verbs.dm:
```diff
@@ -756,294 +327,115 @@ GLOBAL_LIST_INIT(view_runtimes_verbs, list(
-
-	if(!check_rights(R_SERVER | R_EVENT))
-		return
+	to_chat(user, span_interface("You are now a normal player."), confidential=TRUE)
```
```suggestion
	to_chat(user, span_interface("You are now a normal player."), confidential = TRUE)
```

littleboobs replied:
Тот кто вводил confidential, везде копипастил без отступов, надо просто замену через ВСК сделать с 'confidential=TRUE' на 'confidential = TRUE'

dageavtobusnick commented on code/modules/admin/admin_verbs.dm:
```diff
@@ -1119,115 +481,109 @@ GLOBAL_LIST_INIT(view_runtimes_verbs, list(
-		for(var/mob/T as mob in GLOB.mob_list)
-			to_chat(T, chat_box_notice_thick(span_notice("[span_fontsize4("<b>Man up.<br> Deal with it.</b>")]<br>Move on.")))
-			SEND_SOUND(T, sound('sound/voice/manup1.ogg'))
+	if(look_into_inventory)
```
ранний ретурн

dageavtobusnick commented on code/modules/admin/smites.dm:
```diff
@@ -530,21 +530,16 @@
-	set name = "Smite"
-	if(!check_rights(R_EVENT))
+// MARK: Admin smite proc
+ADMIN_VERB_AND_CONTEXT_MENU(admin_smite, R_EVENT, "Smite", "Smite a player with divine power.", ADMIN_CATEGORY_FUN, mob/living/player_mob as mob)
```
тут по идее как и езде нужно добавить in GLOB.player_list

littleboobs replied:
Я если честно разницы и разночтений не понимаю. На Паре везде ин вью и плеер лист, а на ТГ это просто ин ворлд и эс моб. Разницы не то чтобы понял, кроме вью и ворлда. Это микрооптимизация с глоб листом?

dageavtobusnick replied:
Вполне себе не микро. В списке поиск быстрее.

dageavtobusnick commented on code/modules/admin/verbs/BrokenInhands.dm:
```diff
@@ -1,35 +0,0 @@
-		var/F = file("broken_hand_icons.txt")
-		fdel(F)
-		to_chat(F, text)
-		to_chat(world, "Completed and written to [F]")
```
Не знал, что у нас такое вообще есть. Выглядит полезно, только причесать бы.

littleboobs replied:
Я тоже не знал, но это выглядит, как то, что должно быть юнит тестом. Причём код 12 летней давности что ли

dageavtobusnick commented on code/modules/admin/verbs/adminsay.dm:
```diff
@@ -144,20 +118,28 @@
 	BLACKBOX_LOG_ADMIN_VERB("Devsay")
+
+/client/proc/get_dev_team_say()
+	if(check_rights(R_VIEWRUNTIMES|R_ADMIN))
```
ранний ретурн

dageavtobusnick commented on code/modules/admin/verbs/one_click_antag.dm:
```diff
@@ -1,16 +1,8 @@
-		holder.one_click_antag()
-	return
+ADMIN_VERB(one_click_antag, R_SERVER|R_EVENT, "Create Antagonist", "Auto-create an antagonist of your choice.", ADMIN_CATEGORY_EVENTS)
+	if(user.holder)
```
ранний ретурн

littleboobs replied:
bruh

dageavtobusnick commented on code/modules/mob/dead/observer/observer.dm:
```diff
@@ -533,8 +533,55 @@ This is the proc mobs get to turn into a ghost. Forked from ghostize due to comp
 	setDir(2)//reset dir so the right directional sprites show up
 	return ..()
 
+/mob/dead/observer/proc/jump_to_ruin()
```
это по идее тоже админверб

littleboobs replied:
эта шляпа ещё гостами может использоваться, у неё нет лока на пермишены

dageavtobusnick replied:
Она вроде раньше ограничивалась списками вербов

dageavtobusnick committed:
Merge branch 'master220' into admin-verbs

dageavtobusnick committed:
R_HOST fix

dageavtobusnick committed:
request manager fix

littleboobs committed:
bugfixes and dage review

- Traitor Panel открывается из PP
- Logs в PP показывает сикей
- Переписаны на новую систему пропущенные вербы
- Вернул возможность открытия Borg Panel из верб панели, а не только через TP
- Исправлен/улучшен investigate_log
- В TP используются вербы law manager и borg panel вместо создания их тгуи на месте
- Улучшено логирование call shuttle, сплит шаттл вербов в свой файл
- admin_emp, gib_them, admin_explosion и прочее скрыты, как и раньше. Это хуйня для топиков.
- Улучшено логирование некоторых вербов.
- Кодстайл и бла-бла.
- Вернул форс хайджек шаттла

И ещё какие-то фиксы, что я забыл записать.

littleboobs committed:
more hidden topic verbs

littleboobs committed:
rework GENERAL_PROTECT_DATUM

- Изменение GENERAL_PROTECT_DATUM. Он переопределяет флаги datum_protecting_flags в vv проках, что позволяет проверять их на родительском уровне и обходить протект с R_PERMISSIONS. Редактирование datum_protecting_flags через vv запрещено.

littleboobs committed:
ClickOn + CTRL_CLICK show PP again

littleboobs committed:
Launch Supplypod fix

- Launch Supplypod открывался дважды

littleboobs committed:
src --> usr

littleboobs committed:
refactor Send Alert Message

littleboobs committed:
vuap_personal fixes

littleboobs committed:
Merge remote-tracking branch 'upstream/master220' into admin-verbs

littleboobs committed:
A LOT OF FIXES, VUAP fixes and adminjump refactor

- Ликвидированы все рантаймы из-за нулевого клиента при проверке на пермишены.
- Правильное использование динамик инвоуков для вуапа в топике.
- Отрефакторены все проки используемые в джамп ту, джампт ту руин переделан в админ верб с каким-то пермишеном.

littleboobs committed:
ADMIN_STATE

littleboobs committed:
smite_ui fix

littleboobs committed:
last_mind

littleboobs committed:
Fax Panel runtime fix

littleboobs committed:
vuap_personal context menu fix

dageavtobusnick commented on code/modules/admin/veth_player_panel/player_options_panel.dm:
```diff
@@ -1,4 +1,4 @@
-ADMIN_VERB_AND_CONTEXT_MENU(vuap_personal, R_ADMIN, "Open TGUI PP", "Player options panel for a mob.", ADMIN_CATEGORY_GAME, mob/target in GLOB.player_list)
+ADMIN_VERB_AND_CONTEXT_MENU(vuap_personal, R_ADMIN, "Open TGUI PP", "Player options panel for a mob.", ADMIN_CATEGORY_GAME, mob/target in world)
```
Тогда уж GLOB.mob_list

dageavtobusnick commented on code/modules/admin/verbs/playsound.dm:
```diff
@@ -200,12 +200,12 @@ ADMIN_VERB(play_intercomm_sound, R_SOUNDS, "Play Sound via Intercomms", "Plays a
-	if(!M)
-		M = tgui_input_list(user, "Choose a mob to play the sound to. Only they will hear it.", "Play Mob Sound", sort_names(GLOB.player_list))
-	if(!M || QDELETED(M))
+ADMIN_VERB(play_direct_mob_sound, R_SOUNDS, "Play Direct Mob Sound", "Play a sound directly to a mob.", ADMIN_CATEGORY_SOUNDS, sound as sound, mob/target in world)
```
Аналогично

dageavtobusnick commented on code/modules/admin/verbs/adminjump.dm:
```diff
@@ -28,147 +33,124 @@ ADMIN_VERB(jump_to, R_ADMIN, "Jump to...", "Area, Mob, Key or Coordinate", ADMIN
-		return
-
-	var/turf/T = locate(tx, ty, tz)
+ADMIN_VERB_AND_CONTEXT_MENU(jump_to_mob, R_ADMIN, "Jump To Mob", "Jump to any mob in the game.", ADMIN_CATEGORY_GAME, mob/target in world)
```
Аналогично

dageavtobusnick commented on code/modules/admin/verbs/adminjump.dm:
```diff
@@ -28,147 +33,124 @@ ADMIN_VERB(jump_to, R_ADMIN, "Jump to...", "Area, Mob, Key or Coordinate", ADMIN
-	if(isobj(usr.loc))
-		var/obj/O = usr.loc
-		O.force_eject_occupant(usr)
+ADMIN_VERB(jump_to_area, R_ADMIN, "Jump To Area", "Jumps to the specified area.", ADMIN_CATEGORY_GAME, area/target in world)
```
Вроде есть глобальный список со всеми area, если да, то лучше юзать его

dageavtobusnick commented on code/modules/admin/verbs/adminjump.dm:
```diff
@@ -28,147 +33,124 @@ ADMIN_VERB(jump_to, R_ADMIN, "Jump to...", "Area, Mob, Key or Coordinate", ADMIN
-		var/obj/target_loc = target.loc
-		target_loc.force_eject_occupant(target)
-	admin_forcemove(target, get_turf(user.mob))
+ADMIN_VERB_AND_CONTEXT_MENU(get_mob, R_ADMIN, "Get Mob", "Teleport a mob to your location.", ADMIN_CATEGORY_GAME, mob/target in world)
```
Моблист

dageavtobusnick commented on code/modules/mob/dead/observer/observer.dm:
```diff
@@ -557,28 +548,15 @@ This is the proc mobs get to turn into a ghost. Forked from ghostize due to comp
 
-	var/ruinname = tgui_input_list(usr, "Select ruin", "Jump to Ruin", names)
-
+	var/ruinname = input(user, "Select ruin", "Jump to Ruin") as null | anything in sort_list(names)
```
Тгуи инпут лист

littleboobs committed:
dage review 2, fixes and debug_controller

- Новый debug_controller вместо заросшего плесенью бессмысленного debug_misc_controller
- Ревью фиксы, использование глоб листов вместо поиска в мире, пропущенный тгуи инпут.
- Убраны насильное выкидывание админов в гостов при различных джампах, оставлено только у jump_to_key, т.е. по сикею.
- Фикс пермишенов разрабов(???)
- debug_atom_init в админ верб

dageavtobusnick committed:
Merge branch 'master220' into admin-verbs

littleboobs committed:
borg panel

- Использует глоб лист силиконов, улучшено логирование. Добавлены краши. Использование CLIENT_FROM_VAR

littleboobs committed:
inputs and permissions, fixes

- Trigger Event выведен из дебаг флага.
- Скрытие ненужных кнопок, вербы из стат панели используют тгуи инпут листы.
- Менторы не могут смотреть ВВ.
- Фикс пермишенов, деадмин и реадмин, доступы разработчиков и т.д.
- Get Mob работает по кукле.

dageavtobusnick commented on code/modules/admin/holder2.dm:
```diff
@@ -1,34 +1,32 @@
-	var/admincaster_screen = 0	//See newscaster.dm under machinery for a full description
-	var/datum/feed_message/admincaster_feed_message = new /datum/feed_message   //These two will act as holders.
-	var/datum/feed_channel/admincaster_feed_channel = new /datum/feed_channel
-	var/admincaster_signature	//What you'll sign the newsfeeds as
```
Это куда подевалось?

dageavtobusnick commented on code/modules/events/event_procs.dm:
```diff
@@ -1,4 +1,3 @@
-ADMIN_VERB_VISIBILITY(trigger_event, ADMIN_VERB_VISIBLITY_FLAG_MAPPING_DEBUG)
 ADMIN_VERB(trigger_event, R_EVENT, "Trigger Event", "Trigger Event.", ADMIN_CATEGORY_EVENTS, event_type in SSevents.allEvents)
```
event_type in SSevents.allEvents не нужен, тут снизу специально используется инпут лист, а контекстно мы не можем обращаться к датумам.

littleboobs committed:
dage review 3

littleboobs committed:
fixes again

- admin_observe_target (FLW) снова работает для менторов
- Смайт применяется на таргете, а не себе. Смайт работает на мобе, а не клиенте.
- possess_object и release_object снова работают

littleboobs committed:
change_command_name only for R_EVENT

littleboobs committed:
Merge remote-tracking branch 'upstream/master220' into admin-verbs

littleboobs committed:
fix after merge upstream

littleboobs committed:
paper spam permission fix and beauty

littleboobs committed:
ERT and SSD logs

- ЕРТ верб разделён обратно на прок и верб. Динамик инвоуки заменены на проки.
- При открытии логов в ПП передаётся кукла, а не клиент.

littleboobs committed:
get_mob_by_ckey

littleboobs committed:
tgui inputs in proxy verbs

- Тест ареа с ТГ в попапе
- ТГУИ инпуты и хайды для топик кала
и наверное что-то ещё

littleboobs committed:
Logs and ATickets for moderators

littleboobs committed:
booleaningx and 0 -> R_NONE

dageavtobusnick commented on code/modules/admin/verbs/debug.dm:
```diff
@@ -328,20 +328,25 @@ ADMIN_VERB(cmd_debug_make_powernets, R_DEBUG, "Make Powernets", "Regenerates all
 
 ADMIN_VERB_VISIBILITY(cmd_admin_grantfullaccess, ADMIN_VERB_VISIBLITY_FLAG_MAPPING_DEBUG)
-ADMIN_VERB(cmd_admin_grantfullaccess, R_EVENT, "Grant Full Access", "Grant full access to a mob.", ADMIN_CATEGORY_DEBUG, mob/M in GLOB.mob_list)
+ADMIN_VERB(cmd_admin_grantfullaccess, R_EVENT, "Grant Full Access", "Grant full access to a mob.", ADMIN_CATEGORY_DEBUG, mob/target)
```
Вообще это раньше было в контекстном меню. Лучше будет либо контекстнок онли, либо сделать ещё и прокси для статпанели с тгуи.

dageavtobusnick commented on code/modules/admin/verbs/debug.dm:
```diff
@@ -352,89 +357,139 @@ ADMIN_VERB(cmd_admin_grantfullaccess, R_EVENT, "Grant Full Access", "Grant full
-ADMIN_VERB(cmd_assume_direct_control, R_DEBUG|R_ADMIN, "Assume Direct Control", "Assume direct control of a mob.", ADMIN_CATEGORY_DEBUG, mob/M in GLOB.mob_list)
-	if(M.ckey)
-		if(tgui_alert(user, "This mob is being controlled by [M.ckey]. Are you sure you wish to assume control of it? [M.ckey] will be made a ghost.",, list("Yes", "No")) != "Yes")
+ADMIN_VERB(cmd_assume_direct_control, R_DEBUG|R_ADMIN, "Assume Direct Control", "Assume direct control of a mob.", ADMIN_CATEGORY_DEBUG, mob/target in GLOB.mob_list)
```
Это тоже

littleboobs committed:
no more admin_ghost for mentors

dageavtobusnick commented:
/gemini review

littleboobs committed:
Merge remote-tracking branch 'upstream/master220' into admin-verbs

littleboobs commented on code/__HELPERS/text.dm:
```diff
@@ -387,7 +387,7 @@
 	input_text = replacetext(input_text, "\[/u\]", "</u>")
 
-	if(check_rights(R_EVENT))
+	if((findtext(input_text, "\[signfont\]") || findtext(input_text, "\[/signfont\]") && check_rights(R_EVENT))) // Make sure the text is there before giving off an error
```
брэд

littleboobs commented on code/controllers/subsystem/garbage.dm:
```diff
@@ -644,30 +644,15 @@ SUBSYSTEM_DEF(garbage)
-
-	qdel_and_find_ref_if_fail(src, TRUE)
+ADMIN_VERB(qdel_then_if_fail_find_references, R_DEBUG, "qdel() then Find References if GC failure", "qdel() then Find References if GC failure", ADMIN_CATEGORY_DEBUG)
+	qdel_and_find_ref_if_fail(user, TRUE)
```
хз

littleboobs commented on code/datums/mind.dm:
```diff
@@ -2344,15 +2348,9 @@
-				var/datum/borgpanel/B = new(usr, R)
-				B.ui_interact(usr)
-				log_and_message_admins("has opened [R]'s Borg Panel.")
+				SSadmin_verbs.dynamic_invoke_verb(usr, /datum/admin_verb/borg_panel, locateUID(href_list["borgpanel"]))
```
благо

littleboobs commented on code/modules/admin/verbs/randomverbs.dm:
```diff
@@ -1252,44 +1049,33 @@ Traitors and the like can also be revived with the previous role mostly intact.
 			continue
 
-		M.audible_message(span_italics("...ваббаджек...ваббаджек..."))
+		M.audible_message(span_hear("...ваббаджек...ваббаджек..."))
```
так тут и так нужны спан стоит теперь. работай лучше, нейрослоп

littleboobs commented on code/controllers/subsystem/non-firing/admin_verbs.dm:
```diff
@@ -0,0 +1,169 @@
+
+	// refresh their verbs
+	admin_visibility_flags[admin.ckey] ||= list()
+	if(admin.holder.is_localhost_autoadmin) // NOT WORKING!
```
это было намеренно

littleboobs commented:
/gemini review

littleboobs commented on code/__HELPERS/text.dm:
```diff
@@ -387,7 +387,7 @@
 	input_text = replacetext(input_text, "\[/u\]", "</u>")
 
-	if(check_rights(R_EVENT))
+	if((findtext(input_text, "\[signfont\]") || findtext(input_text, "\[/signfont\]") && check_rights(R_EVENT))) // Make sure the text is there before giving off an error
```
тру

littleboobs commented on code/controllers/subsystem/garbage.dm:
```diff
@@ -644,30 +644,15 @@ SUBSYSTEM_DEF(garbage)
-
-	qdel_and_find_ref_if_fail(src, TRUE)
+ADMIN_VERB(qdel_then_if_fail_find_references, R_DEBUG, "qdel() then Find References if GC failure", "qdel() then Find References if GC failure", ADMIN_CATEGORY_DEBUG)
+	qdel_and_find_ref_if_fail(user, TRUE)
```
тру

littleboobs commented on code/game/machinery/computer/communications.dm:
```diff
@@ -334,13 +334,13 @@ GLOBAL_VAR_INIT(captain_auth_access, ACCESS_CAPTAIN)
 				to_chat(ui.user, span_warning("Вашего уровня доступа не хватает для отправки ОБР."))
 				return
-			ui.user.client.send_response_team()// check_rights is handled on the other side, if someone does get ahold of this
+			ui.user.client.send_response_team()
```
это нормально

littleboobs committed:
gemini review and some fixes

- borg_panel fix
- cmd_admin_subtle_message in GLOB.mob_list
- скрыл сломанный admin_observe_target
- ref tracking вербы фикс

littleboobs commented:
/gemini review

littleboobs commented on code/modules/admin/verbs/adminsay.dm:
```diff
@@ -1,140 +1,114 @@
 			continue
 		if(enabling)
-			add_verb(C, msay)
-			to_chat(C, "<b>Mentor chat has been enabled.</b> Use 'msay' to speak in it.")
+			to_chat(client, "<b>Mentor chat has been enabled.</b> Use 'msay' to speak in it.")
 		else
-			remove_verb(C, msay)
-			to_chat(C, "<b>Mentor chat has been disabled.</b>")
+			to_chat(client, "<b>Mentor chat has been disabled.</b>")
 
 	log_and_message_admins("toggled mentor chat [enabling ? "on" : "off"].")
-	BLACKBOX_LOG_ADMIN_VERB("Toggle Msay")
-
-/client/proc/get_dev_team_say()
-	if(check_rights(R_VIEWRUNTIMES | R_ADMIN))
-		var/msg = tgui_input_text(src, null, "devsay \"text\"", encode = FALSE)
-		cmd_dev_say(msg)
-
-/client/proc/cmd_dev_say(msg as text)
-	set name = "Devsay"
-	set hidden = TRUE
-
-	if(!check_rights(R_VIEWRUNTIMES | R_ADMIN)) // Catch any non-admins trying to use this proc
-		return
+	BLACKBOX_LOG_ADMIN_VERB("Toggle MSay")
```
брэд

littleboobs committed:
proxy fix and unit test
